core/matrix/bccoo_aux_structs.hpp

```diff
@@ -0,0 +1,172 @@
+struct compr_idxs {
+    IndexType nblk;  // position in the block
+    IndexType blk;   // active block
+    IndexType row;   // row index
+    IndexType col;   // column index
+    size_type shf;   // shift on the chunk
+
+    GKO_ATTRIBUTES compr_idxs() : nblk(0), blk(0), row(0), col(0), shf(0) {}
+
+    GKO_ATTRIBUTES compr_idxs(IndexType blk_init, size_type shf_init,
+                              IndexType row_init = 0, IndexType col_init = 0,
+                              IndexType nblk_init = 0)
+        : nblk(nblk_init),
+          blk(blk_init),
+          row(row_init),
+          col(col_init),
+          shf(shf_init)
+    {}
+};
+
+
+/**
+ *  Specific struct to manage block compression bccoo objects
+ */
+template <typename IndexType>
+struct compr_blk_idxs {
+    IndexType row_frst;  // minimum row index in a block
+    IndexType col_frst;  // minimum column index in a block
+    IndexType row_diff;  // maximum difference between row indices
+                         // in a block
+    IndexType col_diff;  // maximum difference between column indices
+                         // in a block
+    size_type shf_row;   // shift in chunk where the rows vector starts
+    size_type shf_col;   // shift in chunk where the cols vector starts
+    size_type shf_val;   // shift in chunk where the vals vector starts
+    bool mul_row;        // determines if the block includes elements
+                         // of several rows
+    uint8 rows_cols;     // combination of previous bool variables
+
+    GKO_ATTRIBUTES compr_blk_idxs()
+        : row_frst(0),
+          col_frst(0),
+          row_diff(0),
+          col_diff(0),
+          shf_row(0),
+          shf_col(0),
+          shf_val(0),
+          rows_cols(0)
+    {}
+
+    GKO_ATTRIBUTES compr_blk_idxs(const IndexType* rows_data,
+                                  const IndexType* cols_data,
+                                  const size_type block_size,
+                                  const compr_idxs<IndexType>& idxs,
+                                  const uint8 type_blk)
+        : row_frst(rows_data[idxs.blk]),
+          col_frst(cols_data[idxs.blk]),
+          row_diff(0),
+          col_diff(0),
+          shf_row(idxs.shf),
+          shf_col(idxs.shf),
+          shf_val(idxs.shf),
+          rows_cols(type_blk & type_mask_rows_cols)
+    {
+        shf_col +=
+            ((rows_cols & type_mask_rows_multiple)
+                 ? ((rows_cols & type_mask_rows_16bits) ? sizeof(uint16) : 1) *
+                       block_size
+                 : 0);
+        shf_val = shf_col + block_size * ((rows_cols & type_mask_cols_8bits) ? 1
+                                          : (rows_cols & type_mask_cols_16bits)
+                                              ? sizeof(uint16)
+                                              : sizeof(uint32));
+    }
+
+    GKO_ATTRIBUTES void set_multi_row()
+    {
+        rows_cols |= type_mask_rows_multiple;
```

Comment:
When using block BCCOO, each block is divided into three parts: [ vect_dspl_rows , vect_dspl_cols , vect_vals ].
[ shf_row , shf_col , shf_val ] point to the start of each part, respectively.
In the case, all elements of the block are on the same row, the first part is empty, and the computation of shf_col needs to know this by querying type_mask_rows_multiple.
In the same way, shf_val querys type_mask_cols_8bits and type_mask_cols_16bits to be computed

Reply:
> the computation of shf_col needs to know this by querying type_mask_rows_multiple.

so if call set_multi_row after constructing compr_blk_idxs, it should update shf_col according to this, right?

Reply:
True, I had not considered this situation, and it could be "dangerous" ... 
Since, it is only used once in bccoo_helper, I will copy the expression there. 
I am going to compile it. Thanks!!

Reply:
Works!! Thank you very much!!